src/main/java/teammates/storage/sqlapi/UserDb.java

```diff
@@ -0,0 +1,138 @@
+package teammates.storage.sqlapi;
+
+import java.time.Instant;
+
+import org.hibernate.Session;
+
+import teammates.common.exception.EntityAlreadyExistsException;
+import teammates.common.exception.EntityDoesNotExistException;
+import teammates.common.exception.InvalidParametersException;
+import teammates.common.util.HibernateUtil;
+import teammates.storage.sqlentity.Instructor;
+import teammates.storage.sqlentity.Student;
+import teammates.storage.sqlentity.User;
+
+import jakarta.persistence.criteria.CriteriaBuilder;
+import jakarta.persistence.criteria.CriteriaQuery;
+import jakarta.persistence.criteria.Root;
+
+/**
+ * Handles CRUD operations for users.
+ *
+ * @see User
+ */
+public final class UserDb extends EntitiesDb<User> {
+    private static final UserDb instance = new UserDb();
+
+    private UserDb() {
+        // prevent initialization
+    }
+
+    public static UserDb inst() {
+        return instance;
+    }
+
+    /**
+     * Creates a user.
+     */
+    public <T extends User> T createUser(T user)
+            throws InvalidParametersException, EntityAlreadyExistsException {
+        assert user != null;
+
+        if (!user.isValid()) {
+            throw new InvalidParametersException(user.getInvalidityInfo());
+        }
+
+        if (getInstructor(user.getId()) == null) {
+            throw new EntityAlreadyExistsException(String.format(ERROR_CREATE_ENTITY_ALREADY_EXISTS, user.toString()));
+        }
+
+        if (getStudent(user.getId()) == null) {
+            throw new EntityAlreadyExistsException(String.format(ERROR_CREATE_ENTITY_ALREADY_EXISTS, user.toString()));
+        }
+
+        persist(user);
+        return user;
+    }
+
+    /**
+     * Gets an instructor by its {@code id}.
+     */
+    public Instructor getInstructor(Integer id) {
+        assert id != null;
+
+        return HibernateUtil.getSessionFactory().getCurrentSession().get(Instructor.class, id);
+    }
+
+    /**
+     * Gets a student by its {@code id}.
+     */
+    public Student getStudent(Integer id) {
+        assert id != null;
+
+        return HibernateUtil.getSessionFactory().getCurrentSession().get(Student.class, id);
+    }
+
+    /**
+     * Saves an updated {@code User} to the db.
+     */
+    public <T extends User> T updateUser(T user)
+            throws InvalidParametersException, EntityDoesNotExistException {
+        assert user != null;
+
+        if (!user.isValid()) {
+            throw new InvalidParametersException(user.getInvalidityInfo());
+        }
+
+        if (getInstructor(user.getId()) == null) {
+            throw new EntityDoesNotExistException(ERROR_UPDATE_NON_EXISTENT);
+        }
+
+        if (getStudent(user.getId()) == null) {
+            throw new EntityDoesNotExistException(ERROR_UPDATE_NON_EXISTENT);
+        }
```

Comment:
Existence check should be done by checking that user with id exists, no need individual instructor or student.